raiutils/raiutils/data_processing/data_processing_utils.py

```diff
@@ -0,0 +1,145 @@
+# Copyright (c) Microsoft Corporation
+# Licensed under the MIT License.
+
+import datetime
+import json
+from typing import Any, Dict, List
+
+import numpy as np
+import pandas as pd
+from scipy.sparse import issparse
+from sklearn.utils import check_consistent_length
+
+_DF_COLUMN_BAD_NAME = "DataFrame column names must be strings."\
+    " Name '{0}' is of type {1}"
+_LIST_NONSCALAR = "Lists must be of scalar types"
+_TOO_MANY_DIMS = "Array must have at most two dimensions"
+
+
+def convert_to_list(array, custom_err_msg=None):
+    if issparse(array):
+        if array.shape[1] > 1000:
+            if custom_err_msg is None:
+                raise ValueError("Exceeds maximum number of features for "
+                                 "visualization (1000)")
+            else:
+                raise ValueError(custom_err_msg)
+        return array.toarray().tolist()
+    if isinstance(array, pd.DataFrame):
+        return array.values.tolist()
+    if isinstance(array, pd.Series):
+        return array.values.tolist()
+    if isinstance(array, np.ndarray):
+        return array.tolist()
+    if isinstance(array, pd.Index):
+        return array.tolist()
+    return array
+
+
+def convert_to_string_list_dict(
+        base_name_format: str,
+        ys,
+        sample_array) -> Dict[str, List]:
+    """Convert the given input to a string-list dictionary.
+
+    This function is used to convert arrays in a variety of types
+    into a dictionary mapping column names to regular Python lists
+    (in preparation for JSON serialisation). It is a modification
+    of the feature processing code in :class:`fairlearn.metrics.MetricFrame`.
+
+    The array to be converted is passed in :code:`ys`, and a variety
+    of types are supported. The :code:`sample_array` argument is
+    used in a call to :func:`sklearn.utils.check_consistent_length`
+    to ensure that the resultant lists are of the right length.
+    Finally `base_name_format` is used to generate sequential
+    keys for the dictionary if none are in the supplied :code:`ys`.
+    It must be of the form :code:`'Base String {0}'`, with the
+    :code:`{0}` being replaced by a sequential integer.
+
+    It is not possible to list out all the possible underlying types
+    for :code:`ys`. A brief summary:
+        - :class:`pd.Series`
+        - :class:`pd.DataFrame`
+        - A simple Python list
+        - A Python dictionary with string keys and values which are
+          convertible to lists
+        - Anything convertible to a :class:`np.ndarray`
```

Comment:
missing docs for parameter types/descriptions

Reply:
Done